kolibri/content/utils/channel_import.py

```diff
@@ -0,0 +1,287 @@
+
+        # Use this rather than get_models, as it returns a list of all models, including those
+        # generated by ManyToMany fields, whereas get_models only returns explicitly defined
+        # Model classes
+        self.content_models = list(content_app.models.values())
+
+        # Get the next available tree_id in our database
+        self.tree_id = self.find_unique_tree_id()
+
+    def get_all_destination_tree_ids(self):
+        ContentNodeRecord = self.destination.get_class(ContentNode)
+        return sorted(map(lambda x: x[0], self.destination.session.query(
+            ContentNodeRecord.tree_id).distinct().all()))
+
+    def find_unique_tree_id(self):
+        tree_ids = self.get_all_destination_tree_ids()
+        # If there are no pre-existing tree_ids just escape here and return 1
+        if not tree_ids:
+            return 1
+        if len(tree_ids) == 1:
+            if tree_ids[0] == 1:
+                return 2
+            return 1
+
+        # Do a binary search to find the lowest unused tree_id
+        def find_hole_in_list(ids):
+            last = len(ids) - 1
+            middle = int(last/2 + 1)
+            # Check if the lower half of ids has a hole in it
+            if ids[middle] - ids[0] != middle:
+                # List is only two ids, so hole must be between them
+                if middle == 1:
+                    return ids[0] + 1
+                return find_hole_in_list(ids[:middle])
+            # Otherwise check if there is a hole in the second half
+            if ids[last] - ids[middle] != last - middle:
+                # Second half is only two ids so hole must be between them
+                if last - middle == 1:
+                    return ids[middle] + 1
+                return find_hole_in_list(ids[middle:])
+            # We should only reach this point in the first iteration, if there are no holes in either
+            # the first or the last half of the list, therefore, we just take the max of the list plus 1
+            # Because the list is already sorted, we can just take the last value
+            return ids[-1] + 1
+        return find_hole_in_list(tree_ids)
+
+    def generate_row_mapper(self, mappings=None):
+        if mappings is None:
+            mappings = {}
+
+        def mapper(record, column):
+            if column in mappings:
+                col_map = mappings.get(column)
+                if hasattr(record, col_map):
+                    return getattr(record, col_map)
+                elif hasattr(self, col_map):
+                    return getattr(self, col_map)(record)
+                else:
+                    raise AttributeError('Column mapping specified but no valid column name or method found')
+            else:
+                return getattr(record, column, None)
+        return mapper
+
+    def base_table_mapper(self, SourceRecord):
+        for record in self.source.session.query(SourceRecord).all():
+            yield record
```

Comment:
No, it's not even different from just doing:
`return self.source.session.query(SourceRecord).all()` as that is itself an iterable.